### PR TITLE
Make graphdrivers work with pluginv2.

### PR DESCRIPTION
**Pull request by anusha-ragunathan**
As part of making graphdrivers support pluginv2, a PluginGetter
interface was necessary for cleaner separation and avoiding import
cycles.

This commit creates a PluginGetter interface and makes pluginStore
implement it. Then the pluginStore object is created in the daemon
(rather than by the plugin manager) and passed to plugin init as
well as to the different subsystems (eg. graphdrivers, volumedrivers).
A side effect of this change was that some code was moved out of
experimental. This is good, since plugin support will be stable soon.

Signed-off-by: Anusha Ragunathan anusha@docker.com

### CODE REVIEW

File: plugin/store/store_experimental.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+// Handle sets a callback for a given capability. It is only used by network
+// and ipam drivers during plugin registration. The callback registers the
+// driver with the subsystem (network, ipam).
+func (ps Store) Handle(capability string, callback func(string, *plugins.Client)) {
+	pluginType := fmt.Sprintf("docker.%s/1", strings.ToLower(capability))
+
+	store := &ps
+
+	// Register callback with new plugin model.
+	store.handlers[pluginType] = callback
```
Comment (AkihiroSuda) [inline plugin/store/store_experimental.go]: store.Lock()?

File: plugin/store/store_experimental.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+	if err != nil {
+		logrus.Debugf("Error in json.Marshal: %v", err)
+		return err
+	}
+	ioutils.AtomicWriteFile(ps.plugindb, jsonData, 0600)
+	return nil
+}
+
+// Get returns a plugin matching the given name and capability.
+func (ps Store) Get(name, capability string) (getter.CompatPlugin, error) {
```
Comment (AkihiroSuda) [inline plugin/store/store_experimental.go]: Just question: why not use a pointer receiver for ps?

File: plugin/v2/plugin_experimental.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+			Type:        mount.Type,
+			Options:     mount.Options,
+		}
+		// TODO: if nil, then it's required and user didn't set it
+		if mount.Source != nil {
+			m.Source = *mount.Source
+		}
+		if m.Source != "" && m.Type == "bind" {
+			fi, err := os.Lstat(filepath.Join(rootfs, string(os.PathSeparator), m.Destination)) // TODO: followsymlinks
+			if err != nil {
```
Comment (AkihiroSuda) [inline plugin/v2/plugin_experimental.go]: You don't need to explicitly specify the path separator here?

File: plugin/v2/plugin_experimental.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
+// +build experimental
```
Comment (AkihiroSuda) [inline plugin/v2/plugin_experimental.go]: !windows ?

File: plugin/store/store_experimental.go (hunk 2 above)
Comment (anusha-ragunathan) [inline plugin/store/store_experimental.go]: Done.

File: plugin/store/store_experimental.go (hunk 1 above)
Comment (anusha-ragunathan) [inline plugin/store/store_experimental.go]: Done

File: plugin/v2/plugin_experimental.go (hunk 3 above)
Comment (anusha-ragunathan) [inline plugin/v2/plugin_experimental.go]: Done.

File: plugin/v2/plugin_experimental.go (hunk 4 above)
Comment (anusha-ragunathan) [inline plugin/v2/plugin_experimental.go]: Builds fine on windows, although code path will not be reached. So its okay.

**Comment by anusha-ragunathan** [comment]
Windows CI is failing due to known issues.

**Comment by anusha-ragunathan** [comment]
@cpuguy83 @vieux : PTAL

File: plugin/distribution/pull.go [hunk 5]
```diff
@@ -62,14 +62,26 @@ func (pd *pullData) Layer() (io.ReadCloser, error) {
 	return rsc, nil
 }
 
-// Pull downloads the plugin from Store
-func Pull(name string, rs registry.Service, metaheader http.Header, authConfig *types.AuthConfig) (PullData, error) {
+// GetRef returns the distribution reference for a given name.
+func GetRef(name string) (reference.Named, error) {
```
Comment (vieux) [inline plugin/distribution/pull.go]: I'm not sure I understand the need for this `GetRef` function, `reference.ParseNamed` seems identical

Comment (anusha-ragunathan) [inline plugin/distribution/pull.go]: All imports of `package reference` to be in one place. plugin/distribution.

File: volume/store/store_test.go [hunk 6]
```diff
@@ -5,11 +5,15 @@ import (
 	"strings"
 	"testing"
 
+	pluginstore "github.com/docker/docker/plugin/store"
 	"github.com/docker/docker/volume/drivers"
 	volumetestutils "github.com/docker/docker/volume/testutils"
 )
 
 func TestCreate(t *testing.T) {
+	pluginStore := pluginstore.NewStore("/var/lib/docker")
```
Comment (cpuguy83) [inline volume/store/store_test.go]: tempdir or is this not writing anything?

File: plugin/getter/interface.go [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package getter
+
+import "github.com/docker/docker/pkg/plugins"
+
+// CompatPlugin is a abstraction to handle both v2(new) and v1(legacy) plugins.
+type CompatPlugin interface {
+	Client() *plugins.Client
+	Name() string
+	IsLegacy() bool
```
Comment (cpuguy83) [inline plugin/getter/interface.go]: `IsV1()`?

**Review by cpuguy83 — APPROVED** [review]
Couple of minor things, otherwise LGTM

**Comment by vieux** [comment]
@anusha-ragunathan please rebase :)

File: volume/store/store_test.go (hunk 6 above)
Comment (anusha-ragunathan) [inline volume/store/store_test.go]: It's the rootdir where plugin rootfs and other meta data are created. Its the same as dockerd's graphRoot.

File: plugin/getter/interface.go (hunk 7 above)
Comment (anusha-ragunathan) [inline plugin/getter/interface.go]: Updated.

**Comment by anusha-ragunathan** [comment]
@vieux : rebased. PTAL

File: plugin/distribution/pull.go (hunk 5 above)
Comment (vieux) [inline plugin/distribution/pull.go]: ok thanks

**Comment by vieux** [comment]
LGTM

**Pull request merged**